aip/general/0129.md

```diff
@@ -0,0 +1,150 @@
+
+Server-modified and default values often result in a poor user experience for
+[state-driven clients](state-driven-clients). These clients are often unable to
+tell when their desired state matches the current state for these fields, as the
+rules by which a server may modify and return values are complex, not public,
+and not repeatable. This AIP aims to clarify what behaviors can be expected from
+services and describing how they can be repeated on the client-side.
+
+## Guidance
+
+### Single Owner Fields
+
+Fields should have a single owner, whether that is the client or the server.
+Server owned fields should be indicated with the `OUTPUT_ONLY` field_behavior.
+All other types of fields should be considered to be owned by the client. The
+server should respect the value (or lack thereof) for all client owned fields
+and not modify them. Failure to respect client set values can lead to
+[Declarative clients][] entering an infinite loop to correct the change.
+
+### Effective Values
+
+There are instances where a service will allocate, generate, or calculate a
+value if the client chooses not to specify one. For example: a client may
+create a virtual machine without specifying a static IP address for the virtual
+machine to be available on. Such a scenario is opting into dynamic IP address
+allocation.
+
+Some examples of these types of fields are ones that are:
+
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with an effective value ***must*** be expressed as two fields in
+the API:
+* a mutable field that can be optionally set by the user and **must not** be
+modified by the service
+* an `OUTPUT_ONLY` field that records the effective value decided on by the
+service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string effective_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Naming
+
+Effective values **must** be named by prefixing `effective_` to the mutable
+field's name.
+
+### User-Specified Fields
+
+For user-specified fields, the value in response from the service **must** be
+the same as provided by the create or update request. For string fields this
+means returning the value unchanged, with two exceptions: 
+
+* When a string field has a data type annotation, a normalized string that
+represents the given value may be returned.
+* When a string field has a default value indicated by the `default` keyword,
+that default value may be returned when the given value is empty or missing.
+
+### Normalizations
+
+Only a set of allowed normalizations may be applied. Values that are not annoted
+with the `google.api.value_type` annotation **must not** be normalized in any
+way.
+
+#### Allowed Normalizations
+
+This list of allowed normalizations is defined as follows:
+
+* uuid
+* ipv4
+* ipv6
+* email
+
+Normalizations on fields **must** be described using the `google.api.value_type`
```

Comment:
I think this is the only contentious point left in this PR: the rest looks good.

We can't really introduce guidance without an annotation that goes with it. We could introduce the annotation easily enough: should we just do that? 

This will of course mean that clients will not support these normalizations, at least for a while. But that's probably ok: declarative clients could probably add support for these first, with other normalizations to follow as needed?

Reply:
Let's start drafting an annotation! My two cents would be:
* let's make it of type `message` so that we can extend it in the future if necessary
* let's use an enum for the supported formats

```
message FieldValue {
  enum Format {
    FORMAT_UNSPECIFIED = 0;

    IPV4 = 1;
    IPV6 = 2;
    EMAIL = 3;
    UUID4 = 4;
  }

  Format format = 1;
}

extend proto2.FieldOptions {
  // Example:
  // (google.api.field_value).format = IPV4
  FieldValue field_value = xxxxx;
}
```

Reply:
After the API Design team discussed this, I volunteered to draft an annotation, [take a look](http://cl/557929054) (fyi internal link)

Reply:
I think once that goes in, let's update this PR with some examples.

Reply:
The annotation is now submitted internally, but isn't published yet. This AIP can be updated to mention it, but it should be left in `draft` until we have "launched" the annotation.